torch_lr_finder/lr_finder.py

```diff
@@ -538,7 +538,8 @@ def plot(
                 min_grad_idx = (np.gradient(np.array(losses))).argmin()
             except ValueError:
                 print(
-                    "Failed to compute the gradients, there might not be enough points."
+                    "Failed to compute the gradients, there might not be enough points. "
+                    "Please check whether num_iter >= (skip_start + skip_end + 2)."
                 )
```

Comment:
What do you think of:
-  raising an error if `num_iter < (skip_start + skip_end + 2)` on line ~518:
```py
# Make sure there are enough data points to suggest a learning rate
if suggest_lr and len(lrs) < (skip_start + skip_end + 2):
    raise RuntimeError(
        f"Need at least {skip_start + skip_end + 2} iterations to suggest a"
        f"learning rate. Got {len(lrs)}"
    )
```
- keep this message like before (since now we raise error if there are not enough iterations) but use `warning` instead of print
```py
warnings.warn(
    "Failed to compute the gradients, there might not be enough points."
)
```

Reply:
> > raising an error if num_iter < (skip_start + skip_end + 2) on line ~518
> 
> Yeah, it's better to raise the error earlier. I was intended to keep the behavior same as before by not interrupting the program when this error occurs. But it requires users remember to check the unpacked LR manually, and it's not a straightforward approach for users who are not familiar with the APIs.
> 
> If we all agree on this, I can add this check right after this block of code:
> 
> https://github.com/davidtvs/pytorch-lr-finder/blob/fd9e949bc709e31881c77a4b0710745524c5091e/torch_lr_finder/lr_finder.py#L515-L522

👍 , that would be a good place to put it

> 
> > keep this message like before (since now we raise error if there are not enough iterations) but use warning instead of print
> 
> Regarding this one, I need your opinion. I've roughly checked the implementation of `np.gradient()` from v1.9.1 to v2.1 (the latest one), it seems there is no other `ValueError` would raise in our use case except the one indicating insufficient data points to calculate gradient.
> 
>     * v1.9.0: https://github.com/numpy/numpy/blob/v1.9.0/numpy/lib/function_base.py#L975-L978
>       The only `ValueError` would be raised for insufficient data points.
> 
>     * v1.9.1: https://github.com/numpy/numpy/blob/v1.9.1/numpy/lib/function_base.py#L950-L951
>       A new `ValueError` for an invalid value of the new introduced keyword argument `edge_order`.
> 
>     * v2.1.0: https://github.com/numpy/numpy/blob/v2.1.0/numpy/lib/_function_base_impl.py#L1235-L1239
>       2 new `ValueError`s for ill-formed input arrays (introduced in v1.11.0, I just checked with this revision to see whether there are more `ValueError`s).
> 
> 
> Do you think we can just remove this warning message or even unwrap the `np.gradient()` call from the try-catch block if we decided to raise the error mentioned above earlier? My considerations about this are:
> 
>     1. In our use case, the `ValueError` defined in numpy v1.9.0 is exactly what we want to catch. And it seems not possible to trigger other `ValueError`s introduced in later releases.
> 
>     2. If there is really any specific usage can trigger other `ValueError`s, the existing [try-catch block](https://github.com/davidtvs/pytorch-lr-finder/blob/fd9e949/torch_lr_finder/lr_finder.py#L537-L542) will suppress the actual cause. And I don't think this is good.
> 
>     3. As we decided to raise the error for insufficient data points, user can fix the problem and run it again. Then this warning message is not possible to show anymore.

Makes sense to just remove it. Thanks for checking!

Reply:
Thanks, will submit the update later!

Reply:
Ohhhhh, I misread the message. According to the if-condition, we can directly put the statement right after the first 3 if statements:
https://github.com/davidtvs/pytorch-lr-finder/blob/fd9e949bc709e31881c77a4b0710745524c5091e/torch_lr_finder/lr_finder.py#L506-L511

Anyway, you can check with it after the update is submitted, thanks!

Reply:
Test case is also updated to catch the expected error now.
Feel free to let me know if there is anything need to revise.